ccmbioinfo/lfs
# CARDS-2263: Store HERACLES S3 data export task performance metrics in the JCR

This _Pull Request_ introduces 3 new performance metrics, all related to HERACLES S3 data exports, namely:

- `S3ExportedForms` counts the number of _Forms_ successfully pushed to the configured S3 bucket
- `S3ExportedSubjects` counts the number of Subjects successfully pushed to the configured S3 bucket
- `S3ExportFailures` counts the number of S3 export tasks that have terminated in error

Additionally, an `ErrorTracking` service is made available so that whenever `ErrorTracking.logError` is called, an `nt:file` node in the JCR under `/LoggedEvents` is created which holds the contents of the logged error. The `SlackNotificationsTask` is modified so that any logged errors under `/LoggedEvents` are sent in the periodic Slack alerts update along with the performance metrics recorded under `/Metrics`. As of the current implementation, all logged errors under `/LoggedEvents` will be sent to Slack, therefore once the errors have been resolved, these node under `/LoggedEvents` should be manually deleted so that alerts of these (now resolved) errors are no longer sent to Slack.

As the errors collected by `ErrorTracking.logError` can potentially be long, I needed to switch from using basic _text_ objects on the Slack API to using _attachment_ objects. This automatically adds a _Show more / Show less_ button for messages that would be too long to be sent to Slack using basic _text_ objects. This, however, slightly changes the appearance of the message of all our Slack alert messages.

For example, a nightly performance metric update message from _Cards4PREMs_ would now look like this,

![screenshot_2023_07_12-15-30-19](https://github.com/data-team-uhn/cards/assets/5349728/a2baba55-f1da-4032-8a05-8dc1f83b1af8)

instead of this,

![screenshot_2023_07_12-15-30-32](https://github.com/data-team-uhn/cards/assets/5349728/5224aa0d-9fd1-49a7-bb48-d1bb8a4d434b)

Testing Instructions
--------------------------

1. Build this (`CARDS-2263`) branch including a Docker image with `mvn clean install -Pdocker`.
2. `cd compose-cluster`
3. `python3 generate_compose_yaml.py --dev_docker_image --oak_filesystem --composum --cards_project cards4heracles --slack_notifications --s3_test_container`.
4. You will be prompted to enter a Slack webhook URL. Either create your own Slack channel and app with an _incoming webhook_ or DM me to be added to the _#cards-alerts-dev_ Slack channel and I will provide you with an incoming webhook.
5. Edit `docker-compose.yml` and add the environment variable `NIGHTLY_SLACK_NOTIFICATIONS_SCHEDULE=0 * * ? * *` to the `cardsinitial` container.
6. `docker-compose build && docker-compose up -d`
7. Every minute, a performance metric update containing **Number Of Forms Exported to S3 bucket**, **Number Of Subjects Exported to S3 bucket**, and **Number Of Failed S3 export jobs** should be sent to Slack.
8. Visit http://localhost:8080 and login as `admin`:`admin`
9. Create a _Demographics_ Form and a test _Patient_ Subject.
10. Visit http://localhost:8080/Subjects.s3push?dateLowerBound={TODAY}&dateUpperBound={TOMORROW}. For example, if today is July 12th 2023, you would visit http://localhost:8080/Subjects.s3push?dateLowerBound=2023-07-12&dateUpperBound=2023-07-13.
11. The next update in Slack should show **Number Of Failed S3 export jobs** incremented by 1 followed by a stack trace showing a `NoSuchBucket` error.
12. Visit http://localhost:8080/bin/browser.html and delete the `nt:file` node under `/LoggedEvents`
13. Visit http://localhost:9001 and login as `minioadmin`:`minioadmin`
14. Click on _Create bucket_ and create a new bucket called `uhn`.
15. Visit the `Subjects.s3push` URL once again (eg. http://localhost:8080/Subjects.s3push?dateLowerBound=2023-07-12&dateUpperBound=2023-07-13)
16. The next Slack update should increment both **Number Of Forms Exported to S3 bucket** and **Number Of Subjects Exported to S3 bucket** by 1.
17. Create a new _Baseline Health Information_ Form for the same test patient and visit the S3 push URL (eg. http://localhost:8080/Subjects.s3push?dateLowerBound=2023-07-12&dateUpperBound=2023-07-13) once again.
18. The next Slack update should increment **Number Of Forms Exported to S3 bucket** by 2 and **Number Of Subjects Exported to S3 bucket** by 1.
19. Visit http://localhost:8080/bin/browser.html and delete everything under `/Metrics`.
20. The next Slack update should show _**ERROR:** Could not gather any performance statistics_.
21. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
